hoomd/device.py

```diff
@@ -316,30 +316,33 @@ def __init__(self, nthreads=None, communicator=None, msg_file=None, shared_msg_f
                                                            self.cpp_msg)
 
 
-class Auto(_device):
+def auto(communicator=None, msg_file=None, shared_msg_file=None, notice_level=2):
     """
     Allow simulation hardware to be chosen automatically by HOOMD-blue
 
     Args:
-        nthreads (int): number of TBB threads
         communicator (:py:mod:`hoomd.comm.Communicator`): MPI communicator object. Can be left None if using a
             default MPI communicator
         msg_file (str): Name of file to write messages to
         shared_msg_file (str): (MPI only) Name of shared file to write message to (append partition #)
         notice_level (int): Minimum level of notice messages to print
 
-    TODO: convert this to a function that produces a GPU or CPU device.
-    """
-
-    def __init__(self, nthreads=None, communicator=None, msg_file=None, shared_msg_file=None, notice_level=2):
-
-        _device.__init__(self, communicator, notice_level, msg_file, shared_msg_file)
 
-        _init_nthreads(nthreads)
+    Returns:
+        a GPU or CPU device, depending on availability, GPU is preferred
+    """
 
-        self.cpp_exec_conf = _hoomd.ExecutionConfiguration(_hoomd.ExecutionConfiguration.executionMode.AUTO,
+    self=_device(communicator, notice_level, msg_file, shared_msg_file)
+    self.cpp_exec_conf = _hoomd.ExecutionConfiguration(_hoomd.ExecutionConfiguration.executionMode.AUTO,
                                                            [],
                                                            False,
                                                            False,
                                                            self.comm.cpp_mpi_conf,
                                                            self.cpp_msg)
+    
+    # Set class according to C++ object
+    if self.cpp_exec_conf.isCUDAEnabled():
+        self.__class__=GPU
```

Comment:
I was surprised as well, but yes, it actually does work. As for better methods, I'm happy with any changes. We also felt this was ugly, but didn't see a better way to leave all logic on both sides of the code unchanged.

Reply:
My first thought is why not just

````python
try:
    device = hoomd.device.GPU(args)
except SomeException:
    device = hoomd.device.CPU(args)
````

It seems to me this would work, and have the desired behavior. I know we have the auto option in the `ExecutiionConfiguration`, but does it do more than this would? Python has an idea that it is better to ask forgiveness than permission.

If we don't want to do that, then we could add classmethods and change `_device.__init__` to take an `ExecutionConfiguration`, but it will be more complicated.

Reply:
`ExecutionConfiguration` gives verbose errors when the requested configuration is not available. The auto-selection logic in C++ is already robust so I'd prefer to use it. Alternately, we could refactor that logic so that it can be called from Python. I don't know if that would be more or less complicated than a classmethod based solution.

Reply:
@joaander That was actually our first approach. To expose getNumCapableGPUs, however making it callable from before the `ExecutionConfiguration` object exists is what that didn't resolve and thus it still had the logic of create first and modify after. So I ultimately simplified it to using the already existing method `isCudaEnabled`.

Reply:
Are these errors pass through as exceptions using pybind? If so then the verbosity shouldn't be a problem. I am just trying to see the gain from an auto option other than masking the error messages.

The problem is we create the execution configuration in the `GPU` and `CPU` constructors. We can't get around needed to call `__init__` cleanly eventually. This means we cannot construct a `ExecutionConfiguration` object in `def auto` and pass it to the correct class. We could give `_device` the responsibility to create the object and then create classmethods for `GPU` and `CPU` that take in an execution configuration and pass it to `device` which could optionally accept an execution environment. That is the best other solution, I can think of. Since `_device` is private it can have a messy constructor, and take in private classes.

Reply:
Yeah, if GPU and CPU took optional `ExecutionConfiguration` objects was one other thing. But it seemed confusing for most users to add a specific keyword argument and I wasn't sure about adding a `**kwarg` in this case although I think that could be pythonic.

Reply:
> Are these errors pass through as exceptions using pybind? If so then the verbosity shouldn't be a problem. I am just trying to see the gain from an auto option other than masking the error messages.

I don't believe they are. I only see a runtime_error and a message about being built without GPU support. I think we'd need to make changes creating more errors to make the exceptions more specific.

Reply:
I would be in favor of making that logic available in python; the implementation of `auto()` then becomes very clean. It would look something like:
```
if _hoomd.getNumCapableGPUs() > 0:
    device = GPU(params)
else:
    device = CPU(params)

return device
```

Reply:
@j-proc pybind11 automatically converts `std::runteme_error` into `RuntimeError`. Also, I wouldn't want to pollute the `GPU` and `CPU` constructors with a private class keyword argument. I think we would want to move the construction of the configuration to the `_device` class. @tommy-waltmann 's idea could work.

Reply:
I was referring to all the noise printed by the Messenger in scanGPUs() and other places to give users explicit messages regarding why specific GPUs were not selected for execution. This is often essential information for the user - e.g. to know that their driver is out of date and that is why HOOMD isn't running on the GPU.

If we refactor ExecutionConfiguration at the C++ level we should include clear ways to get this type of diagnostic information.

It is unfortunate that Python makes multiple constructors for classes so difficult.

Reply:
In order to refactor the C++ logic, one could make the `scanGPUs` method static so it could be called from outside the class, and have it take extra arguments to account for the instance variables that are used inside right now. When it gets called from outside the class, give it replacements for the instance variables that are used now and then use those to return the correct value from the function that will be exported to python.